### DISCUSSION TITLE
Helping you get started with new API references

### DISCUSSION THREAD

**Discussion by jpmedley** [mdn/mdn-community#573] [Content]
Gang,

Writing new API reference requires a bit of set up. Folders and files need to be created in a specific structure. Standard headings and metadata need to be filled in.

To remove friction from this process, I've created [a repo that contains a bunch of boilerplates](https://github.com/jpmedley/mdn-boilerplates) for content that doesn't exist. All you need to do is:

1. Create a new branch in your clone of mdn/content.
1. In the boilerplates repo, locate the folder for the API you want to document. 
1. Copy it or download it to your new mdn/content branch.
1. Add text and examples to the boilerplates.
1. Commit your new content and submit a pull request as normal.

If the API you want to document is missing from the repo, contact jmedley@google.com. 

If you have reason to think someone else might want to work on a feature, [use this process](https://github.com/jpmedley/mdn-boilerplates#getting-started). Otherwise have at it. (I'm looking for the simplest, least bureaucratic way to prevent duplication of error. Let me know if you have an idea.)

**Comment by hamishwillee** [comment]
I was thinking of creating a yari template wizard to make things like this "generically". This is at least as good and possibly better. Does your generator  know about method prototypes so that the syntax blocks can be auto expanded out?

On spot check, the syntax does not quite match what we're doing now for JavaScript: `DOMStringList.contains()` would be `contains()`.

**Reply by jpmedley** [reply]
Do you mean, does it create syntax section in the boilerplates? Yes. If you mean something else, please clarify.

Regarding the problem with `contains()`, that is the fault of [my tool's templates](https://github.com/jpmedley/mdn-helper/tree/main/templates) being out of date. These will be fixed in my next update of the boilerplates.

**Reply by jpmedley** [reply]
I just pushed an update. The syntax problem is fixed.

**Reply by hamishwillee** [reply]
Thanks very much! Yes, that was my problem.

@Elchi3 you might want to check out some of these. For example I don't think https://github.com/jpmedley/mdn-boilerplates/blob/main/api/applicationcache/oncached/index.md matches our intent for the syntax, but I could be wrong.

What I like about this over a user-run tool is that the list of things to do is clear, and assuming this is up to date and correct, hard to screw up.

Some ignorable thoughts
- https://github.com/jpmedley/mdn-boilerplates/tree/main/api/mutationevent_landing is a landing page. Should mutationevent have a landing page? Should it be named as landing page - after all most landing pages are named something like mutations_api
- The landing page already on MDN for this is https://developer.mozilla.org/en-US/docs/Web/API/MutationEvent at https://github.com/mdn/content/blob/main/files/en-us/web/api/mutationevent/index.md . 
  -  I see this is not in your tree in https://github.com/jpmedley/mdn-boilerplates/tree/main/api/mutationevent - does this mean things that are already in MDN are not generated?
  - Could/should landing pages provide auto list of links to the sub pages?
- I wonder if the templates could auto populate the structure of parameters and return value section with at least the names of the params in a definition list?

**Reply by jpmedley** [reply]
> @Elchi3 you might want to check out some of these. For example I don't think https://github.com/jpmedley/mdn-boilerplates/blob/main/api/applicationcache/oncached/index.md matches our intent for the syntax, but I could be wrong.

These are built from templates. It takes me a matter of minutes to edit the templates and rerun the boilerplates. Just let me know what to change.

>https://github.com/jpmedley/mdn-boilerplates/tree/main/api/mutationevent_landing is a landing page. Should mutationevent have a landing page? Should it be named as landing page - after all most landing pages are named something like mutations_api

My intention was not to have any landing pages in the repo. (This is a bug I will fix.) The relationship between landing pages and interfaces is 1 to many. The composition of the many requires human judgement. My intention was not to automate this. 

That said, if someone were to decided the landing pages that outstanding work _should_ be on, I could generate landing pages just for those.

>I see this is not in your tree in https://github.com/jpmedley/mdn-boilerplates/tree/main/api/mutationevent - does this mean things that are already in MDN are not generated?

Yes. If my build tool finds a method like Widget.doStuff(), it constructs a URL like `https://developer.mozilla.org/en-US/docs/Web/API/Widget/doStuff` and pings it. If I get back 200, I don't generate a boilerplate. False negatives are possible, but I currently have my retry count at 5, so I think they're unlikely.

>Could/should landing pages provide auto list of links to the sub pages?

Yes, but as stated above, a human will need to decided the relationships. This is because landing pages don't always have a 1:1 relationship with a spec. Even if there were, I'm not set up to consume specs. Either way, I need to write code. It's on my list.

>I wonder if the templates could auto populate the structure of parameters and return value section with at least the names of the params in a definition list?

Great idea. I need to write code. It's on the list. What do you want first, the parameter lists or populated landing pages?

**Reply by jpmedley** [reply]
I'm also working on code to prepopulate interfaces pages with the interface's members. I know this seems like something I would have had on day one. There's a product history reason for it. I won't bore you with details.

**Reply by jpmedley** [reply]
I've removed the landing pages until we can figure out which ones we need specifically.

**Reply by hamishwillee** [reply]
Interface members would have been my answer, that's going to make things easier. I'd like @Elchi3 to look at the syntax sections because he has a better grasp on this.

Next time I need to create a page I'll use one of yours and might have more feedback. 

At some point we could perhaps consider using this to replace some of the heavily duplicated and over-explained stuff you reported in https://github.com/mdn/content/issues/7980

**Reply by jpmedley** [reply]
>Next time I need to create a page I'll use one of yours and might have more feedback.

I will do my best to respond quickly (as Ruth and Rachel can attest). To manage expectations, know that some requested changes will be easier to implement than others. 

Where is interface members located?

**Reply by jpmedley** [reply]
>Interface members would have been my answer, that's going to make things easier.

Where is this located?

**Reply by hamishwillee** [reply]
@jpmedley 

> Interface members would have been my answer, that's going to make things easier.

Sorry, missed that. Should learn to read.

We might be talking at cross purposes. What I have been thinking about is auto-populating methods and parameters in API reference pages - i.e. in the templates: https://developer.mozilla.org/en-US/docs/MDN/Structures/Page_types/API_reference_page_template#methods

**Reply by jpmedley** [reply]
I had to write some new code. I'm resolving the last bugs and hope to have updated reference pages uploaded later today.

**Reply by jpmedley** [reply]
Reference pages are now pre-populated with methods, parameters, and events. I plan to do something similar with method arguments, but I don't have an ETA.

**Reply by hamishwillee** [reply]
Nice! I think that saves a bit of time - examples https://github.com/jpmedley/mdn-boilerplates/blob/main/api/domfilesystem/index.md and https://github.com/jpmedley/mdn-boilerplates/blob/main/api/animationeffecttimingreadonly/index.md

Something weird going on here: https://github.com/jpmedley/mdn-boilerplates/blob/main/api/canvascapturemediastreamtrack/index.md#syntax

There was a discussion somewhere about generating this sort of thing from Yari. FYI I pointed here.

**Reply by jpmedley** [reply]
>Something weird going on here:

There's a typo in the IDL. Look for a fix in the next push.

>There was a discussion somewhere about generating this sort of thing from Yari. FYI I pointed here.

Any possibility you could focus on CSS? What I've done so far is not trivial. I've been trying to get to CSS for 3 years, but can't find the time.

**Reply by hamishwillee** [reply]
> Any possibility you could focus on CSS? 

Unfortunately not. I'm weak on CSS and only do that when the rest of the team runs out of bandwidth. i.e. I don't think I'd add much value.

**Comment by pepelsbey** [comment]
Closing as outdated/irrelevant. Let me know if that’s not the case.
